### ISSUE TITLE
The sidebar is stretching to the width of the threads list content

### ISSUE THREAD

**Issue by auge8472** [My-Little-Forum/mylittleforum#740]
With the layout changes in version 20240827.1 the main content of the overview pages are grids. This causes a stretching of the width of the grid in case of long subject lines in the threads list. The stretched grid causes a stretching of the sidebar to the then available width.

![mlf2-stretched-sidebar](https://github.com/user-attachments/assets/c8892ca5-2c29-436b-8b58-f30f354a3edc)

Before this change (prior to version 20240827.1) the sidebar stood on its own in narrow viewports and was a floating box in wider viewports. I don't want to go back to this. It would make it necessary to have one definition for narrow viewports and a completely different definition for wider viewports instead to have one definition (the grid) with slightly different changes for the different viewport widths. Because of this I will perform tests with fixed widths of the grid and with its elements (including the sidebar) and their overflow behaviour.

**Comment by auge8472** [comment]
I've made a few tests. The reason for the "over"-sized sidebar is `grid-template-columns: 1fr;` for the grid, which makes it taking the necessary width, that is normally the width of the thread list, that takes often wore width than the narrow viewports provide. This also extends the sidebar to this width.

There are two solutions.

**First one** is to restrict the width of the sidebar to match the viewport width (including margins, paddings and border widths).

~~~css
#sidebar {
  grid-area: sidebar;
  position: relative;
  margin: 0;
  padding: 0;
  /*
  the next restricts only the sidebar, it depends
  on the value of the body padding */
  max-width: calc(100vw - 2em);
}
~~~

![Bildschirmfoto zu 2024-09-18 13-00-30](https://github.com/user-attachments/assets/465ebbe0-e770-4fe1-b4c9-ca2889700205)

The thread list itself exceeds the viewport width (if needed) and makes horizontal scrolling necessary.

**Second possible solution** is to replace the width value for the grid itself in narrow viewports.

~~~css
#main-grid {
  display: grid;
  /*grid-template-columns: 1fr;*/
  grid-template-columns: calc(100vw - 2em);
  grid-template-rows: auto auto;
  gap: .75em;
  grid-template-areas:
"sidebar   " 
"threadlist";
}
~~~

![Bildschirmfoto zu 2024-09-18 13-10-05](https://github.com/user-attachments/assets/dae6f184-3de1-4fd6-8b5e-47420648ae49)

This restricts not only the width of the sidebar but also of the thread list to the available viewport width. This was (achieved with other CSS properties) the look and behaviour between the versions 20220529.1 and 20240729.1. No need for horizontal scrolling but more vertical scrolling because of the possible multiline nature of a list entry for a single posting. This can be seen, if you compare the display of the thread tree in the two screenshots.

I would prefer the second solution but I am open to counter-proposals.

**Comment by loesler** [comment]
I agree. +1 for the second solution.

**Comment by joeiacoponi1** [comment]
Auge, I'd like to ask a favor before chiming in. Can you display a thread that has additional indentation sublevels / subposts? The thread you show here has a total of 3 levels (initial plus 2 sublevels).  With this displayed thread, the second solution looks and works far nicer than the first solution.

However, using a 20240827 copy of our production forum as an example, our very first thread actually goes to 8 levels (7 sublevels), and causes the most indented posts to only have one word per line. I don't honestly know what the maximum sublevel usage is on our board, but I suspect it is far closer to 20 than even 8. With our current production board at 2.4.24, a lot of horizontal scrolling is required, but at least the post titles are more readable, similar to the first solution.

I'm honestly not sure there is a perfect answer here, as it probably depends on how a given board is typically used, but I wanted to make sure you were at least aware of our challenges.

**Comment by auge8472** [comment]
@joeiacoponi1 I created a thread in my testing forum with a maximum depth of 8 replies in one of the branches but I was not able to replicate the view with the non wrapped lines (like in the screenshot of my first solution proposal).

Even if the line breaks do not match the viewport width, they are still present and occur as soon as possible. As a result, the subjects can also be word-wrapped if the depth is appropriate.

**Figure 1**, wrapping with solution 1 (width set for sidebar (viewport width - 2em)):

![Screenshot of a thread tree, the width was set for the sidebar above](https://github.com/user-attachments/assets/88a80bfa-4a92-409d-9a9c-e60393933ef9)

One can see the end of the sidebar in the first two pixel lines of the image. The entry subject lines does not fit into the viewport width of 420 pixels but it wraps at the next possible gap (at any separator character (i.e. a space or `-`)).

**Figure 2**, wrapping with solution 2 (width set for the whole grid element including the sidebar (viewport width - 2em)):

![Screenshot of a thread tree, the width was set for the grid, that includes the sidebar above](https://github.com/user-attachments/assets/07f61da6-b9e0-4304-864a-81260ad29301)

Also here the end of the sidebar is visible on top of the image. The entry subject lines break earlier, according to the given width.

**Figure 3**, Wrapping behaviour of higher indentation levels *in **both** proposed solutions*

![Screenshot of a paet of a thread tree with a higher indentation level](https://github.com/user-attachments/assets/4c496dcf-b658-4e4a-9174-bae628898c11)

In both solutions the subjects of postings with a higher indetation level are wrapped, possibly behind every single word.

I was not able to replicate the behavior, seen in the first screenshot of [my previous comment](https://github.com/My-Little-Forum/mylittleforum/issues/740#issuecomment-2358222656) with non breaking lines in the thread tree. This might be caused by different browser configurations. All screenshots in this issue were made with Firefox or Librewolf in versions 129 and 130 with activated developer tools and "test browser sizes" tool (with a viewport width set to 420px) but on different computers with different OS's (both Windows and Linux).

**Comment by joeiacoponi1** [comment]
Thanks Auge.  

For reference, here are some screenshots from our forum, both current production (2.4.24) and latest (20240827).  [Tested on Safari with iPhone 13 Pro / iOS 18]

MLF 2.4.24 - Recent forum thread, from OP to 7 levels deep.
![IMG_4166](https://github.com/user-attachments/assets/474bcd65-8d08-4530-bccd-351ead5d58f8)

MLF 20240827 - Same recent thread as above. but doesn't have all the posts. Good for perspective.
![IMG_4167](https://github.com/user-attachments/assets/88c2d274-8f2e-4746-86fa-8be9e3029ef2)

MLF: 20240827 - This particular thread goes to 15 levels. The first is a screenshot of an older thread, leading up to the following 2 images. The next two are side-by-side screenshots of the same thread beginning at around 10 levels deep, and the photos are of the same spot with the second one scrolled to the right. 
![IMG_4168](https://github.com/user-attachments/assets/f89dadec-fe19-4bf3-be9b-1b4313ec2dbc)
![IMG_4163](https://github.com/user-attachments/assets/418fff96-5991-40b5-862e-82746223b66a)
![IMG_4164](https://github.com/user-attachments/assets/353bc000-9dd4-4504-8297-e39e77fdb9cf)

Review & Thoughts:
I'll preface this all by saying I understand we may be an outlier in certain instances (forum size, # of sublevels), but the board has been running strong for 15+ years and over 500,000 posts. 

MLF 2.4.24: In mobile, our production forum is pretty difficult to read unless you zoom the screen. Requires a lot of side to side scrolling plus additional zooming.

MLF 20240827: While not perfect for our forum in cases where sublevels gets to 8+ levels, it seems to strike a fair balance. I do like how the text size is decreased as the sublevel increases. It allows for more words on the line in those instances. Much better than 2.4. While not the most aesthetically beautiful compared to your proposals, 20240827 is the most functional version. One potential change tp 20240827 would be to decrease the indentation amount on each level, but this doesn't really help to get us where you want to go with a clean look to the board.

Proposals 1 and 2: Given that they react the same at higher indentation levels, I'll discuss them the same here.  With your latest post to 8 levels, you can probably see how it becomes a challenge for us. While it does eliminate horizontal scrolling, the titles become pretty difficult to read.

Conclusion: A perfect implementation might be one where the display is Proposal 2, up to the point where the # of sublevels on the given pages reaches a threshold, but that seems extremely unwieldy to calculate / implement. I haven't yet come across any solution that can handle all these scenarios well, and feel like the best approach might be a forum setting which allow each board Admin to choose either the layout of 20240827 or that of Proposal 2.  I don't know how complex that would be, however.

**Comment by auge8472** [comment]
> I understand we may be an outlier in certain instances (forum size, # of sublevels)

I don't think so (at least for the number of sublevels/reply-levels). Both the project forum as well as some other forums contain threads with a deep reply-level structure. I have just had a look at the (MLF1 based) forum of Alfie (one of our project forum moderators). It's about pharmaceutical development and contains several threads with many reply-levels (I actually found a thread with 18 reply-levels). He adapted several changes to the HTML-structure to make his forum mobile aware. Deep threads in his forum need horizontally scrolling. He capped the length of an entries list entry with making the subject an inline block element, what forces a linebreak between the subject and the metadata of an entry (author, time of creation and functional links, if present). The subject lines and metadata in itself does not wrap inside in his forum.

![Screenshot of a forced line break in a thread tree entry of a posting](https://github.com/user-attachments/assets/f6048da9-381d-4cc5-a83d-7d07cc37379f)

This gave me an idea. What, if I set `text-wrap: nowrap;` for the subject in itself.

**Figure 1**, non breaking subject with width set for the grid (see previous comments for implementation details):

![Screenshot of a thread with non breaking subject lines, width set for the grid](https://github.com/user-attachments/assets/e0540a71-9223-4fce-8ae7-97f0fd15ee37)

The subject lines do not break inside but because of a suboptimal HTML-structure, there are several breaks behind the subjects.

**Figure 2**, non breaking subject with width set for the sidebar (see previous comments for implementation details):

![Screenshot of a thread with non breaking subject lines, width set for the sidebar](https://github.com/user-attachments/assets/94d72d1a-b0a2-40ea-a22f-068cf5398534)

Screenshot of the thread with the same postings as in figure 1. The line breaks sometimes occur in other places. The metadata are displayed together because there is no need for a line break inside. To keep i.e. the subject and the name as well as all metadata and the functional links together, we would need a rework of the HTML-structure. Something like the following.

~~~html
<li>
 <span class="subject-line">
  <a href="...">This is the subject</a> - <strong>Authors name</strong>
 <span>
 <span class="metadata">
  <time datetime="1970-01-01 00:00:01">01.01.0970, 00:00:01</time>
  <a title="preview">...</a>
  <a title="open whole thread">...</a>
  <a title="and">...</a>
  <a title="so">...</a>
  <a title="on">...</a>
 </span>
</li>
~~~

~~~css
ul.thread li .subject-line,
ul.thread li .metadata {
  text-wrap: nowrap;
}
/* identify the authors name to display a comma after it */
ul.thread li .subject-line strong::after {
  content(", ");
}
~~~

*This is just a first draft.* With something like this we could ensure the breaking point to be predictable and consistent over all entries, *that needs the line break*. Maybe the authors name should move to the metadata. 🤔

---

> MLF 2.4.24: In mobile, our production forum is pretty difficult to read unless you zoom the screen. ...

The forums default theme of the version 2.4.24 is in general not mobile aware. Because of this, your description and screenshot takes no wonder. I can't imagine wanting to read anything especially in the tag list. 😅

---

> MLF 20240827: While not perfect ... it seems to strike a fair balance.

Yes, it's (not surprisingly) far more better, even with the giant looking font size. 🙂

> I do like how the text size is decreased as the sublevel increases.

This is (more or less) a bug because I forgot to adapt the general changes for the thread tree elements to the elements with the additional classes `.deep-reply` and `.very-deep-reply`, what leads us to the next point.

> Two potential changes to 20240827 would be 1) to decrease the left indentation amount on each level and 2) decrease the white space on the right in the titles.

At the moment the indentation size is a complex construct of negative marging and positive padding for the reply lists (`ul`) and margins for their list items (`li`) to display the arrows and the dotted connecting lines between the branches in a correct way. Apart from the problem of the complexity of the current implementation it would be possible to implement a constantly decreasing indentation with margin and padding values smaller than 1 with a relative unit like `em` (i.e. `0.9em`). However, this would be at the expense of the graphical representation with arrows and connecting lines as it currently exists. At some point, the images would no longer match the indentation values.

>  A perfect implementation might be one where the display is Proposal 2, up to the point where the # of sublevels on the given pages reaches a threshold, but that seems extremely unwieldy to calculate / implement.

This is generally implemented (forum settings `deep_reply` and `very_deep_reply` with their associated classes). **But** these settings was originally meant to restrict the indentation in desktop browsers (mobile devices was no thing at that time) and can store only one value each. This makes them optimally usable for either mobile view **or** desktop view **but not both**. If one set i.e. `deep_reply` to the value of `5`, which is possibly an optimal value for displaying the thread tree in narrow viewports, it could look odd in a desktop browser with a far more wide viewport. Conversely, a high value for deep_reply (i.e. `20`) would make no sense, when viewing the forum in a mobile device with a narrow viewport.

There is a saying in german language, "Einen Tod musst du sterben" (litarally "You have to die a death"). I have found a few equivalents for the English language. It's something like "You can't have your cake and eat it too." or "Something's got to give" (I don't know, if one of those fits perfectly). Means, you have to choose between a possibly odd view in different browsers or horizontally scrolling (if necessary). The latter is possible without restriction.

> I feel like the best approach might be a forum setting which allow each board Admin to choose either the layout of 20240827 or that of Proposal 2.

Therefore we have the theming support. But that would make it necessary to adapt changes in later versions. This is not for forum operators with no knowledge of HTML and CSS. It can be circumvented by overwriting styling properties behind the point when the stylesheet is loaded in the main template.

~~~html
<head>
 <!-- document header elements -->
 <link rel="stylesheet" type="text/css" href="[https://mylittleforum.net/forum/themes/default/style.min.css](view-source:https://mylittleforum.net/forum/themes/default/style.min.css)" media="all" />
 <!-- further document header elements -->
 <style>
 /* overwrite the rulesets from inside the stylesheet file */
 </style>
~~~

This would be a way for overwriting *one or only a few* stylesheet definitions.

Far better way is to provide a best as possible solution by default.

**Comment by joeiacoponi1** [comment]
> This is just a first draft. With something like this we could ensure the breaking point to be predictable and consistent over all entries, that needs the line break. Maybe the authors name should move to the metadata. 🤔

This is brilliant. Given that the subject is limited to 60 characters, you already ensure it won't be massively wide on its own.  I also love the idea of moving the author's name into metadata, and putting all metadata into a nowrap.  That would ensure if there is a line break that the author shows up at the beginning of the second line, which seems perfect.
![Screenshot 2024-09-20 at 7 32 01 AM](https://github.com/user-attachments/assets/e33fd168-1a86-47d3-928b-ea909027d038)

> This is (more or less) a bug because I forgot to adapt the general changes for the thread tree elements to the elements with the additional classes .deep-reply and .very-deep-reply, what leads us to the next point.

Heh, well I'm not opposed to it! I would definitely go with the above ideas first, and only re-implement if the width was still too unwieldy.

And noted on your indentation and theming support.  

Overall, I really do like the simplicity of the nowrap combined with moving the author into metadata also in a nowrap.  It should eliminate horizontal scrolling in most instances, which sounds like a win.

**Comment by auge8472** [comment]
Ok, I changed the HTML-structure to introduce a grouping element for the metadata. I removed the string ` - ` between the subject and the authors name from the HTML source and replaced it with a space in the HTML source and the CSS-rule `.metadata::before { content: " - "; }`.

Adding …

~~~css
ul.thread li .subject-line,
ul.thread li .metadata {
  text-wrap: nowrap;
}
~~~

… resulted in …

![Screenshot of a thread tree in a narrow viewport with a typographical separator between the subject and the authors name, added before the authors name](https://github.com/user-attachments/assets/4df54a39-37d1-4306-9004-ffe712c2b809)

The ` - ` disturbed me in a narrow viewport. It doesn't look right at the start of a line. Because of this I removed the rule in the browsers developer tool. This resulted in …

![Screenshot of a thread tree in a narrow viewport without a typographical separator between the subject and the authors name](https://github.com/user-attachments/assets/1cb392dc-775a-4477-927e-cb9f2ee8500f)

Looks good to me. **But** in a wider screen with non breaking lines it looks a bit odd now.

![Screenshot of a thread tree in a wide viewport without a typographical separator between the subject and the authors name](https://github.com/user-attachments/assets/7bc2b475-6a26-43e4-80ea-8f64563638dd)

There one can see, that a separator is needed, IMHO the space character is not enough to mark the gap between the subject and the autors name. With the ` - ` string it looks far better but IMHO also not right when there is a line break between subject and metadata (as in the narrow view). I fiddled a bit around and moved the string to the end of the subject link. I tried it with a comma (`, `) but that lloks weird behind a subject, that ends with a punctuation mark. So the ` - ` behind the subject looks IMHO best.

![Screenshot of a thread tree in a narrow viewport with a typographical separator between the subject and the authors name, added after the subject](https://github.com/user-attachments/assets/1847a33f-2867-4b4d-8cf9-c62dbc07876c)

In this screenshot, taken in a wide viewport with one non breaking and a breaking line, both possible cases are shown.

**Comment by auge8472** [comment]
@loesler Even I opted for version 2 at the beginning, I did it with the screenshot 2 of [my first comment](https://github.com/My-Little-Forum/mylittleforum/issues/740#issuecomment-2358222656) in mind. But reminding me about threads with a deep reply-level-structure, I think now, the threads need the room for taking more horizontal space.

Splitting the line behind the subject to restrict the length of a line, looks like a good compromise between the need of horizontal scrolling and the taken vertical space for one item. What do you think?

**Comment by loesler** [comment]
Works for me, too!

**Comment by auge8472** [comment]
> Works for me, too!

@loesler I will check, if other places needs an adaption of the changes. If ready I will create a pull request for it.

**Comment by joeiacoponi1** [comment]
@auge8472, both options look great. My personal vote would be for the dash (-) leading the second line, primarily to identify it is a continuation of the line before.  With no dash on the 2nd line, I see a new viewer unsure whether the post starts with the title or user name, but that would diminish with familiarity.  Good job.

**Comment by auge8472** [comment]
First things first: @joeiacoponi1 I decided to place the dash behind the subject. *For me* the leading dash in front of the metadata looks a bit odd and additionaly the dash behind the subject simulates a hyphen within a word in the case of a line break. For me, this also fits better conceptually.

This said, I assume to have found every place where thread trees are displayed. These are the threaded main view and the thread tree below the posting content in the single posting, that are working similar. I adapted the changes I made to both views and it's working. Then I came across the thread trees in the table based overview page.

![Screenshot of a thread in the table based overview page, dash added after the subject with ::after](https://github.com/user-attachments/assets/8e2ff052-d8d4-4939-a321-8dd25f2f39b4)

The CSS-selector `ul.thread li > a::after` selects also the subject in this view even there is no metadata and the dashes are orphaned in front of the function links. With the solution, that @joeiacoponi1 prefers, this would not happen because without the metadata element there is also nothing to select with a `::before` for this element. Searching a while for a way to exclude the subject from the rule, if it is inside the element `table#threadlist`. Today I learned how to apply the combination of the CSS pseudoclasses `:not` and `:has`. Et voilà, with the selector `main:not(:has(table#threadlist)) ul.thread li > a::after` we select the subject only, if it is *not* a descendant of the table `table#threadlist` which in itself is a descendant of `main`. This is, what is called "syntactically sugar" and could possibly be the aim of simplifications, but in principle it works.

![Screenshot of a thread in the table based overview page, no dash added because of the exclusion of the ancestor element table](https://github.com/user-attachments/assets/0e064b55-fd83-4590-b359-9f05e5a2ea2a)


This syntax works in Safari since version 15.4 (released 2022-03-14), Chromium based browsers since version 105 (released 2022-08-30) and Firefox based browsers since version 121 (released 2023-12-19).

It's possible to restrict the rule to browsers with support for `:has` (which is the one used selector with the shortest time of browser support).

```css
@supports(selector(:has(p))) {
  /* add the selector and its rule(s) here */
}
```

I am not sure, if this is necessary because older browser versions without support for `:has` should ignore the rule anyway. And really old browsers doesn't support `@supports` anyway which makes this IMHO useless.
